### DISCUSSION TITLE
uvx ruff check vscode problems parsing

### DISCUSSION THREAD

**Discussion by EtienneT** [astral-sh/ruff#14152] [Q&A]
riff is amazing, but in vscode it only finds problem in opened files which is fine for performance.  But it would be nice to setup a vscode task where you call `uvx ruff check` and it gets parsed in the Problems tab with the problemMatcher.  This way you can call ruff check on your whole code base, and have a list of all problems (including files that are not opened) in the Problems tab.  You can do something similar with eslint in javascript which is very handy.  Right now the python problemMatcher doesn't seem to match the output of ruff check.

```
{
    // See https://go.microsoft.com/fwlink/?LinkId=733558
    // for the documentation about the tasks.json format
    "version": "2.0.0",
    "tasks": [
        {
            "label": "ruff check",
            "type": "shell",
            "command": "uvx ruff check",
            "problemMatcher": [
                "$python"
            ]
        }
    ]
}
```

**Comment by dhruvmanila** [comment]
I think what you're referring to is [workspace diagnostics](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#workspace_diagnostic) which is a planned feature for the red knot language server. This would be similar to how the Python extension provides an option in VS Code called `python.analysis.diagnosticMode`:

```jsonc
  // Analysis scope for showing diagnostics.
  //  - openFilesOnly: Analyzes and reports errors on only open files.
  //  - workspace: Analyzes and reports errors on all files in the workspace.
  "python.analysis.diagnosticMode": "openFilesOnly",
```

Regarding your query, I'm not sure what the expected output format for the `$python` "problemMatcher" is. We do provide multiple output format which can be specified using the `--output-format` command-line flag (refer to `ruff check --help` for details).

**Comment by manekovskiy** [comment]
@EtienneT You can define a problem matcher manually using a regular expression. The default output format can be parsed with the following:

```
...
"problemMatcher": [
	{
		"owner": "Ruff",
		"pattern": [
			{
				"regexp": "^(.*?):(\\d+):(\\d+): (.*)$",
				"file": 1,
				"line": 2,
				"column": 3,
				"message": 4
			}
		]
	}
]
...
```
